### ISSUE TITLE
TRecapcha version 2

### ISSUE THREAD

**Issue by zendre4** [pradosoft/prado#560]
Hello, 

The control TRecaptcha is in version 1 of ReCaptcha.  
Would it be possible to implement the version 2 of ReCaptcha ?

Best wishes 

Julien Stalder

![unnamed](https://cloud.githubusercontent.com/assets/9673982/12047488/a977bec2-aecc-11e5-8868-6594ac36e3fe.gif)

**Comment by david0** [comment]
It would be nice to have support for ReCaptcha 2. A pull request for implementing this is welcome.

I guess it would be the best to implement this in a new control called `TReCaptcha2`.
(for backward compatibility and maintainability)

**Comment by camilohaze** [comment]
[TReCaptcha2.txt](https://github.com/pradosoft/prado/files/81268/TReCaptcha2.txt)

JS Control TReCaptcha2

Prado.WebUI.TReCaptcha2 = jQuery.klass(Prado.WebUI.Control,
{
    onInit: function(options)
	{
		var element = options.ID;
		var options = options.data;

		if (options['callback'] !== undefined)
			options['callback'] = eval(options['callback']);

		if (options['expired-callback'] !== undefined)
			options['expired-callback'] = eval(options['expired-callback']);

		grecaptcha.render(element, options);
	},

	render: function(widget_id, args)
	{
		
	},

	reset: function(widget_id)
	{
	},

	response: function(widget_id)
	{
		
	},

	success: function(response)
	{
		$('#g-recaptcha-response').text(response).change();
		$('#g-recaptcha-response').val(response).change();
	}
});

JS TReCaptcha2Validator

Prado.WebUI.TReCaptcha2Validator = jQuery.klass(Prado.WebUI.TBaseValidator,
{
	onInit : function()
	{
		var obj = this;
		var elements = document.getElementsByName(this.options.ResponseFieldName);
		if (elements)
		if (elements.length>=1)
		{
			this.observe(elements[0],'change',function() { obj.responseChanged() });
			this.observe(elements[0],'keydown',function() { obj.responseChanged() });
		}
	},

	responseChanged: function()
	{
		var field = jQuery("#" + this.options.ID+'_1').get(0);
		if (field.value=='1') return;
		field.value = '1';
		Prado.Validation.validateControl(this.options.ID);
	},

	/**
	 * Evaluate validation state
	 * @function {boolean} ?
	 * @return True if the captcha has validate, False otherwise.
	 */
	evaluateIsValid : function()
	{
		return (jQuery("#" + this.options.ID+'_1').get(0).value=='1');
	}
});

**Comment by ctrlaltca** [comment]
@camilohaze nice, do you want to open a PR with it?

**Comment by camilohaze** [comment]
@ctrlaltca So far it lacks some features such as render methods, and getResponse reset. In the next few hours will post another more stable.

**Comment by camilohaze** [comment]
´´´php
<?php

Prado::using('System.Web.UI.ActiveControls.TCallbackEventParameter');
Prado::using('System.Web.UI.ActiveControls.TActivePanel');

class TReCaptcha2 extends TActivePanel implements ICallbackEventHandler, IValidatable
{
	const ChallengeFieldName = 'g-recaptcha-response';
	private $_isValid=true;
	private $_googleUrl="https://www.google.com/recaptcha/api/siteverify";

    public function __construct()
    {
        parent::__construct();
        $this->setAdapter(new TActiveControlAdapter($this));
    }
    public function getActiveControl()
    {
        return $this->getAdapter()->getBaseActiveControl();
    }
    public function getClientSide()
    {
        return $this->getAdapter()->getBaseActiveControl()->getClientSide();
    }
	public function getClientClassName()
	{
		return 'Prado.WebUI.TReCaptcha2';
	}
	public function getTagName()
	{
		return 'div';
	}
	/**
	 * Returns true if this control validated successfully. 
	 * Defaults to true.
	 * @return bool wether this control validated successfully.
	 */
	public function getIsValid()
	{
		return $this->_isValid;
	}
	/**
	 * @param bool wether this control is valid.
	 */
	public function setIsValid($value)
	{
		$this->_isValid=TPropertyValue::ensureBoolean($value);
	}
	public function getValidationPropertyValue()
	{
		return $this->Request[$this->getResponseFieldName()];
	}
	public function getResponseValue()
	{
		return $this->Request[$this->getResponseFieldName()];
	}
	public function getResponseFieldName()
	{
		return self::ChallengeFieldName;
	}
	/**
	 * Returns your site key. 
	 * @return string.
	 */
	public function getSiteKey()
	{
		return $this->getViewState('SiteKey');
	}
	/**
	 * @param string your site key.
	 */
	public function setSiteKey($value)
	{
		return $this->setViewState('SiteKey', TPropertyValue::ensureString($value));
	}
	/**
	 * Returns your secret key. 
	 * @return string.
	 */
	public function getSecretKey()
	{
		return $this->getViewState('SecretKey');
	}
	/**
	 * @param string your secret key.
	 */
	public function setSecretKey($value)
	{
		return $this->setViewState('SecretKey', TPropertyValue::ensureString($value));
	}
	/**
	 * Returns your language.
	 * @return string.
	 */
	public function getLanguage()
	{
		return $this->getViewState('Language', 'en');
	}
	/**
	 * @param string your language.
	 */
	public function setLanguage($value)
	{
		return $this->setViewState('Language', TPropertyValue::ensureString($value), 'en');
	}
	/**
	 * Returns the color theme of the widget. 
	 * @return string.
	 */
	public function getTheme()
	{
		return $this->getViewState('Theme', 'light');
	}
	/**
	 * The color theme of the widget.
	 * Default: light
	 * @param string the color theme of the widget.
	 */
	public function setTheme($value)
	{
		return $this->setViewState('Theme', TPropertyValue::ensureString($value), 'light');
	}
	/**
	 * Returns the type of CAPTCHA to serve.
	 * @return string.
	 */
	public function getType()
	{
		return $this->getViewState('Type', 'image');
	}
	/**
	 * The type of CAPTCHA to serve.
	 * Default: image
	 * @param string the type of CAPTCHA to serve.
	 */
	public function setType($value)
	{
		return $this->setViewState('Type', TPropertyValue::ensureString($value), 'image');
	}
	/**
	 * Returns the size of the widget.
	 * @return string.
	 */
	public function getSize()
	{
		return $this->getViewState('Size', 'normal');
	}
	/**
	 * The size of the widget.
	 * Default: normal
	 * @param string the size of the widget.
	 */
	public function setSize($value)
	{
		return $this->setViewState('Size', TPropertyValue::ensureString($value), 'normal');
	}
	/**
	 * Returns the tabindex of the widget and challenge.
	 * If other elements in your page use tabindex, it should be set to make user navigation easier.
	 * @return string.
	 */
	public function getTabIndex()
	{
		return $this->getViewState('TabIndex', 0);
	}
	/**
	 * The tabindex of the widget and challenge.
	 * If other elements in your page use tabindex, it should be set to make user navigation easier.
	 * Default: 0
	 * @param string the tabindex of the widget and challenge.
	 */
	public function setTabIndex($value)
	{
		return $this->setViewState('TabIndex', TPropertyValue::ensureInteger($value), 0);
	}
	/**
	 * Renders the container as a reCAPTCHA widget and returns the ID of the newly created widget.
	 * - container: The HTML element to render the reCAPTCHA widget.
	 *   Specify either the ID of the container (string) or the DOM element itself.
	 * - parameters: An object containing parameters as key=value pairs, for example, {"sitekey": "your_site_key", "theme": "light"}.
	 *   See grecaptcha.render parameters. https://developers.google.com/recaptcha/docs/display#render_param
	 */
	public function build($siteKey, $args)
	{
		// Pending
	}

	/**
	 * Resets the reCAPTCHA widget.
	 * Optional widget ID, defaults to the first widget created if unspecified.
	 */
	public function reset($widgetId = null)
	{
		// Pending
	}

	/**
	 * Gets the response for the reCAPTCHA widget.
	 * Optional widget ID, defaults to the first widget created if unspecified.
	 */
	public function response($widgetId = null)
	{
		// Pending
	}

	protected function getClientOptions()
	{
		$options['ID'] = $this->getClientID();
		$options['EventTarget'] = $this->getUniqueID();
		$options['data']['sitekey'] = $this->getSiteKey();
		if ($theme = $this->getTheme()) $options['data']['theme'] = $theme;
		if ($type = $this->getType()) $options['data']['type'] = $type;
		if ($size = $this->getSize()) $options['data']['size'] = $size;
		if ($tabIndex = $this->getTabIndex()) $options['data']['tabindex'] = $tabIndex;

		$options['data']['callback'] = 'this.callback';
		$options['data']['expired-callback'] = 'this.callbackExpired';

        return $options;
    }

    protected function registerClientScript()
    {
		$id			= $this->getClientID();
		$options	= TJavaScript::encode($this->getClientOptions());
		$className	= $this->getClientClassName();
		$cs			= $this->Page->ClientScript;
		$code 		= "var $id = new $className($options);";

		$cs->registerPradoScript('ajax');
		$cs->registerEndScript("grecaptcha:$id", $code);
    }

	public function validate()
	{
		if ($this->Page->IsPostBack)
		{
			if ((is_null($this->getResponseValue())) || (empty($this->getResponseValue())))
				return false;
		}
		else
		{
			if (empty($this->response()))
				return false;
		}

		return true;
	}

	/**
	 * Checks for API keys
	 * @param mixed event parameter
	 */
	public function onPreRender($param)
	{
		parent::onPreRender($param);

		if("" == $this->getSiteKey())
			throw new TConfigurationException('recaptcha_publickey_unknown');
		if("" == $this->getSecretKey())
			throw new TConfigurationException('recaptcha_privatekey_unknown');

		// need to register captcha fields so they will be sent postback
		$this->Page->registerRequiresPostData($this->getResponseFieldName());
		$this->Page->ClientScript->registerHeadScriptFile('grecaptcha2', 'https://www.google.com/recaptcha/api.js?hl=' . $this->getLanguage());
	}

	protected function addAttributesToRender($writer)
	{
		$writer->addAttribute('id',$this->getClientID());
		parent::addAttributesToRender($writer);
	}

    public function raiseCallbackEvent($param)
    {
		$this->onCallback($param);
	}

    public function onCallback($param)
    {
        $this->raiseEvent('OnCallback', $this, $param);
    }

	public function render($writer)
	{
		$this->registerClientScript();
		parent::render($writer);
	}
}

class TReCaptcha2CodeErrors extends TEnumerable
{
	const MissingInputSecret='missing-input-secret';
	const InvalidInputSecret='invalid-input-secret';
	const MissingInputResponse='missing-input-response';
	const InvalidInputResponse='invalid-input-response';
}

Prado::using('System.Web.UI.WebControls.TBaseValidator');

class TReCaptcha2Validator extends TBaseValidator
{
	protected $_isvalid = null;

	protected function getClientClassName()
	{
		return 'Prado.WebUI.TReCaptcha2Validator';
	}
	public function getEnableClientScript()
	{
		return true;
	}
	protected function getCaptchaControl()
	{
		$control = $this->getValidationTarget();
		if (!$control)
			throw new Exception('No target control specified for TReCaptcha2Validator');
		if (!($control instanceof TReCaptcha2))
			throw new Exception('TReCaptcha2Validator only works with TReCaptcha2 controls');
		return $control;
	}
	public function getClientScriptOptions()
	{
		$options = parent::getClientScriptOptions();
		$options['ResponseFieldName'] = $this->getCaptchaControl()->getResponseFieldName();
		return $options;
	}
	/**
	 * This method overrides the parent's implementation.
	 * The validation succeeds if the input control has the same value
	 * as the one displayed in the corresponding RECAPTCHA control.
	 *
	 * @return boolean whether the validation succeeds
	 */
	protected function evaluateIsValid()
	{
		// check validity only once (if trying to evaulate multiple times, all redundant checks would fail)
		if (is_null($this->_isvalid))
		{
			$control = $this->getCaptchaControl();
			$this->_isvalid = $control->validate();
		}
		return ($this->_isvalid==true);
	}
	public function onPreRender($param)
	{
		parent::onPreRender($param);

		$cs = $this->Page->getClientScript();
		$cs->registerPradoScript('validator');

		// communicate validation status to the client side
		$value = $this->_isvalid===false ? '0' : '1';
		$cs->registerHiddenField($this->getClientID().'_1',$value);
		
		// update validator display
		if ($control = $this->getValidationTarget())
		{
			$fn = 'captchaUpdateValidatorStatus_'.$this->getClientID();

			$cs->registerEndScript($this->getClientID().'::validate', implode(' ',array(
				// this function will be used to update the validator
				'function '.$fn.'(valid)',
				'{',
				'  jQuery('.TJavaScript::quoteString('#'.$this->getClientID().'_1').').val(valid);',
				'  Prado.Validation.validateControl('.TJavaScript::quoteString($control->ClientID).'); ',
				'}',
				'',
				// update the validator to the result if we're in a callback 
				// (if we're in initial rendering or a postback then the result will be rendered directly to the page html anyway)
				$this->Page->IsCallback ? $fn.'('.$value.');' : '',
				'',
				// install event handler that clears the validation error when user changes the captcha response field
				'jQuery("#'.$control->getClientID().'").on("change", '.TJavaScript::quoteString('#'.$control->getResponseFieldName()).', function() { ',
					$fn.'("1");',
				'});',
			)));
		}
	}
}
´´´

**Comment by camilohaze** [comment]
The component work :+1:

@ctrlaltca PR.

**Comment by zendre4** [comment]
Thanks for fast response !

Can someone open a pull request?

**Comment by ctrlaltca** [comment]
@camilohaze added, thank you. I had to rework the way the control is rendered, since the build() method was executed before google's js finished loading, resulting in a ```grecaptcha is not defined``` javascript error. Now the build() method is called as a result of google's api on load callback.

**Comment by camilohaze** [comment]
@ctrlaltca Excellent work ... congratulations!

**Comment by yoisar** [comment]
Waw!! Very nice! Excellent!

**Comment by zendre4** [comment]
Tanks for implement it!
